### PR TITLE
Bluetooth: Add connect event structure to zsector

### PR DESCRIPTION
**Pull request by LingaoM**
Add an optional version for connection events
to save some memory.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

### CODE REVIEW

**Comment by jhedberg** [comment]
I'd rather us only have one kind of registration, so I'd remove the old kind of this new kind is introduced. I suppose the only thing you can't do with this new kind is to unregister the callbacks, but do we have any use cases for that?

**Comment by LingaoM** [comment]
> I'd rather us only have one kind of registration, so I'd remove the old kind of this new kind is introduced. I suppose the only thing you can't do with this new kind is to unregister the callbacks, but do we have any use cases for that?

In fact, there is currently no unregister function, and I did not modify this place , because it is not called after the main entry. https://github.com/zephyrproject-rtos/zephyr/blob/0c03a0572bf12ee94b42eb32e820a9788d9731b7/tests/bluetooth/tester/src/gap.c#L1084-L1088

**Comment by LingaoM** [comment]
> I'd rather us only have one kind of registration, so I'd remove the old kind of this new kind is introduced. I suppose the only thing you can't do with this new kind is to unregister the callbacks, but do we have any use cases for that?

@jhedberg Another way is that we can use a macro to control the old way(`bt_conn_cb_register`) and set it to off by default

**Comment by Thalley** [comment]
@LingaoM Do you have data to show the memory saved? 

Like Johan, I'd prefer having only a single way of doing things in the API, so if we want to add this, or change the current way, then I think the memory saving should be noticeable. 

I think this also opens up a myriad of changes, as we probably should add this change to all other callback registers in the Bluetooth stack.

**Comment by LingaoM** [comment]
> @LingaoM Do you have data to show the memory saved?
> 
> Like Johan, I'd prefer having only a single way of doing things in the API, so if we want to add this, or change the current way, then I think the memory saving should be noticeable.
> 
> I think this also opens up a myriad of changes, as we probably should add this change to all other callback registers in the Bluetooth stack.

We can easily calculate the RAM usage:
`sizeof(struct bt_conn_cb) = 20 bytes` (Min)

`CONFIG_BT_SMP=y`
Add another 8 bytes 

`CONFIG_BT_USER_PHY_UPDATE=y`
Add another 4 bytes 

`CONFIG_BT_USER_DATA_LEN_UPDATE=y`
Add another 4 bytes 

`CONFIG_BT_REMOTE_INFO=y`
Add another 4 bytes 

`sizeof(struct bt_conn_cb) = 40 bytes` (Max)

In addition, if we remove the old way, then the list header will not exist,
which will reduce the ram by an additional 4 bytes.

**Comment by jhedberg** [comment]
AFAIK we do have several callbacks that are already handled through linker sections, but I do agree that the goal should be to have consistency throughout the stack. 

I also think @LingaoM has a pretty strong argument here in that the only reason the current conn_cb can't be in ROM is the linked list pointer, so it's not just about shaving off the pointer but moving the entire struct from RAM to ROM. This combined with the fact (which I didn't earlier realize) that we don't have an unregister() function makes me think we could go ahead with this, assuming the old API is removed. The removal of the old API should follow the stable API process, right @carlescufi ?

**Comment by carlescufi** [comment]
> The removal of the old API should follow the stable API process, right @carlescufi ?

For every stable API (and Bluetooth is one of them) there are two ways to modifying it:

- Deprecate the old one, introduce a new one. Maintain the deprecated one for 2 releases, then remove it
- Follow the [Stable API change procedure](https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html#introducing-incompatible-changes) to avoid going through deprecation

So it's one or the other, but it needs to be one of those.

**Comment by joerchan** [comment]
I commented on the RFC before this one.

But in my opinion having the dynamic one in addition (with the addition of an unregister) makes sense.
For example a dynamic GATT service (bt_gatt_service_register) might want to be enabled and disabled, and then use the register and unregister of the conn callbacks in the same way.

**Comment by Thalley** [comment]
> I commented on the RFC before this one.
> 
> But in my opinion having the dynamic one in addition (with the addition of an unregister) makes sense.
> For example a dynamic GATT service (bt_gatt_service_register) might want to be enabled and disabled, and then use the register and unregister of the conn callbacks in the same way.

If I understand correct, you are then in favor of having both (similar to how we actually have both dynamically and statically registered GATT services)?

**Comment by joerchan** [comment]
> If I understand correct, you are then in favor of having both (similar to how we actually have both dynamically and statically registered GATT services)?

Yes, I think that might be useful for service implementations that might want to disable their connection callbacks.
@trond-snekvik Any opinions on this, Mesh Proxy is the only GATT service that does this at the moment.

**Comment by trond-snekvik** [comment]
Dynamic GATT services is a must-have for mesh, but it doesn't unregister the connection callbacks, and we don't really need to. 

I think the callbacks are a little different from the GATT services, as you can just keep a flag in the callback and return early if you're not really interested in processing it at that time. I'm not really against adding it, but if we haven't had any requests for this yet, I'm not sure we really need it.

**Comment by joerchan** [comment]
Well the question is really if we want to remove the existing register function, which would be a shame if we wanted to add the unregister one later.

**Comment by Thalley** [comment]
We use the dynamic GATT service register in LE Audio as well. Mostly because it's near impossible to split primary and secondary services without it, but it also has some other uses. 

Regarding unregistering the callbacks there is an easy workaround which is just updating the pointer of the functions in the callback struct to NULL - It will still be registered, but no function will be called.

**Comment by joerchan** [comment]
Ok, I'm fine with either decision then.

**Comment by LingaoM** [comment]
> Regarding unregistering the callbacks there is an easy workaround which is just updating the pointer of the functions in the callback struct to NULL - It will still be registered, but no function will be called.

If we use the static method (stored in ROM) to replace the existing dynamic method, how to modify the content of this structure?

> as you can just keep a flag in the callback and return early if you're not really interested in processing it at that time

I think this is a feasible way.

**Comment by Thalley** [comment]
> > Regarding unregistering the callbacks there is an easy workaround which is just updating the pointer of the functions in the callback struct to NULL - It will still be registered, but no function will be called.
> 
> If we use the static method (stored in ROM) to replace the existing dynamic method, how to modify the content of this structure?

Setting the callback to NULL would only work with the dynamic registering. Overall as @joerchan pointed out, adding this functionality from this PR is not any different than the dynamic/static way we have of defining GATT services, so I'll retract my comment regarding not wanting to have more than one way of doing it.

**Review by Thalley — APPROVED** [review]
Based on the comments, I think having both a dynamic and static way of registering callbacks is alright.

**Comment by LingaoM** [comment]
> Based on the comments, I think having both a dynamic and static way of registering callbacks is alright.

Can we reconsider this comment? https://github.com/zephyrproject-rtos/zephyr/pull/36119#issuecomment-858516203

**Comment by jhedberg** [comment]
I'm fine having a Kconfig option to enable dynamic connection callbacks (the option would default to disabled) if there are some good known use cases for it. However, it sounds like all existing in-tree users could (and probably will) be converted to the static registration, so I'm not sure it makes sense to keep the old API around after that.

**Comment by LingaoM** [comment]
I have an idea that I don't know how to satisfy:

``` C
struct bt_conn_cb {

#if defined(CONFIG_BT_CONN_CB_FLAG)
        bool *not_linked; /* for example */
#endif

        ......

```

``` C
Z_STRUCT_SECTION_FOREACH() {

#if defined(CONFIG_BT_CONN_CB_FLAG)
        if (cb->not_linked && !(*cb->not_linked)) {
                continue;
        }
#endif

        // call()

        ......
```


``` C
#if defined(CONFIG_BT_CONN_CB_FLAG)
static bool flag = true;
#endif

BT_CONN_CB_DEFINE(conn_callbacks) = {
#if defined(CONFIG_BT_CONN_CB_FLAG)
        .not_linked = &flag,
#endif
	.connected = connected,
	.disconnected = disconnected,
};
```

File: include/bluetooth/conn.h
Comment (jhedberg) [inline include/bluetooth/conn.h:944]: Is this really the common or preferred style for this kind of macros, where you leave it up to the user to provide the struct definition after the macro?

I'd have thought that it's more common to provide the given struct members as parameters to the macro. Also, I think providing a variable name as input is irrelevant since it's not needed anywhere in the code. Instead you can generate unique names using e.g. `__COUNTER__`. So the usage of the macro would then look simply like:
```
BT_CONN_CB_DEFINE(connected, disconnected);
```

Comment (jhedberg) [reply include/bluetooth/conn.h:944]: That said, I do see the advantage of leaving the definition details to the user, since it does produce code that's more readable, and also covers the case you pointed out earlier where you have the need to have some members build-time conditional.

Comment (trond-snekvik) [reply include/bluetooth/conn.h:944]: Relevant conversation: https://github.com/zephyrproject-rtos/zephyr/pull/28054#issuecomment-715168117

**Comment by carlescufi** [comment]
@joerchan @jhedberg could you comment further so we can get this resolved?

**Review by joerchan — APPROVED** [review]
This looks OK to me

> I'm fine having a Kconfig option to enable dynamic connection callbacks (the option would default to disabled)

Currently this is not in, should we do it before merging?

> I have an idea that I don't know can to satisfy:

I didn't understand this one.

**Comment by carlescufi** [comment]
@LingaoM could you please rebase and reply to @joerchan's comments?

**Comment by LingaoM** [comment]
> > I have an idea that I don't know can to satisfy:
> 
> I didn't understand this one.

Sorry for not explaining clearly, but just want to propose a solution.

**Comment by erbr-ot** [comment]
Hi @LingaoM - just a heads up re. https://github.com/zephyrproject-rtos/zephyr/pull/55921 
Would seem the change in this PR is not 100% correct. The refactor will not function correctly when several subtests are compiled into a single image and therefore declare the callback more than once. And I guess it is not caught when the callback handlers are doing (more or less) the same thing...

**Comment by Thalley** [comment]
> Hi @LingaoM - just a heads up re. #55921 Would seem the change in this PR is not 100% correct. The refactor will not function correctly when several subtests are compiled into a single image and therefore declare the callback more than once. And I guess it is not caught when the callback handlers are doing (more or less) the same thing...

Indeed. We've had similar issues in the LE Audio tests, where we needed to do something like we did in https://github.com/zephyrproject-rtos/zephyr/pull/55599. 

But I'd argue that since this is now 2 years old, we can't really blame the changes in this PR for any current test issues :) If your tests don't work if the connection callbacks are registered statically, then you should just revert to use the dynamically register or to only declare one static callback structure @erbr-ot

**Comment by erbr-ot** [comment]
@Thalley I completely agree, just thought I'd give @LingaoM a heads up other than just a comment re. the commit in my revert;-)

**Pull request merged**
